R/best_case_worst_case.R

```diff
@@ -0,0 +1,114 @@
+#' Calculates Best case and Worst case for Emission profile at product level
+#'
+#' @param data Dataframe. Emissions profile product level output
+#'
+#' @return A dataframe
+#' @export
+#'
+#' @examples
+#' library(tiltToyData)
+#' library(readr)
+#' restore <- options(readr.show_col_types = FALSE)
+#'
+#' companies <- read_csv(toy_emissions_profile_any_companies())
+#' co2 <- read_csv(toy_emissions_profile_products_ecoinvent())
+#' europages_companies <- read_csv(toy_europages_companies())
+#' ecoinvent_activities <- read_csv(toy_ecoinvent_activities())
+#' ecoinvent_europages <- read_csv(toy_ecoinvent_europages())
+#' isic_name <- read_csv(toy_isic_name())
+#' profile_emissions_at_product_level <- profile_emissions(
+#'   companies,
+#'   co2,
+#'   europages_companies = europages_companies,
+#'   ecoinvent_activities = ecoinvent_activities,
+#'   ecoinvent_europages = ecoinvent_europages,
+#'   isic = isic_name
+#' ) |>
+#'   unnest_product()
+#'
+#' result <- best_case_worst_case(profile_emissions_at_product_level)
+#' result
+#'
+#' # Cleanup
+#' options(restore)
+best_case_worst_case <- function(data) {
+  crucial_cols <- c(col_companies_id(), col_ep_product(), col_benchmark(), col_emission_profile())
+  check_matches_col_names(data, crucial_cols)
+
+  data |>
+    mutate(
+      n_distinct_products = n_distinct(.data[[col_ep_product()]], na.rm = TRUE),
+      .by = col_companies_id()
+    ) |>
+    mutate(
+      equal_weight = ifelse(.data[[col_n_distinct_products()]] == 0, NA,
+        1 / .data[[col_n_distinct_products()]]
+      )
+    ) |>
+    lowest_risk_category_per_company_benchmark(.by = c(col_companies_id(), col_benchmark())) |>
+    highest_risk_category_per_company_benchmark(.by = c(col_companies_id(), col_benchmark())) |>
+    mutate(
+      best_risk = assign_risk(
+        .data, col_emission_profile(),
+        col_min_risk_category_per_company_benchmark()
+      )
+    ) |>
+    mutate(
+      worst_risk = assign_risk(
+        .data, col_emission_profile(),
+        col_max_risk_category_per_company_benchmark()
```

Comment:
Nice. `col_emissions_profile()` certainly makes this "module" easier to maintain. However it seem that the name should have "risk_category" in it. That's is the more general concept that this column represents. That's also suggested by the fact that it's called together with `col_max_risk_category_per_company_benchmark()` -- which also has "risk_category" in its name. 

Note that the relationship between these columns would be clearer if you their name starts with the same prefix, which also makes them discoverable via autocomplete (i.e. start typing and press TAB). So I propose:

* `col_risk_category_something()`
* `col_risk_category_something_else()`

 BTW, I note that `col_min_risk_category_per_company_benchmark()` seems to be called much less frequently than `col_emission_profile()`, so it's not as important to call it through an interface. It's still nice but the payoff in maintainability is lower than it is for `col_emission_profile()`.

Reply:
@maurolepore Thanks for the info! :)

We have new columns namely `best_risk`, `worst_risk`, `min_risk_category_per_company_benchmark` etc. These new risk columns are used to create `best_case` and `worst_case` outputs. Please note that the `emission_profile` is a risk column too, however this name helps me to create distinction between emission_profile risk and best/worst case risk. Therefore, I propose that we keep the column `col_emissions_profile()` as it is. What do you think?

I would also need to change the function name `assign_risk` as it is not correct and I can't change it to desired name if `col_emission_profile()` becomes `col_risk_category_emission_profile()`. Does it makes sense?

Reply:
Thanks. It'll be easier for me to understand when I see the next round of changes.